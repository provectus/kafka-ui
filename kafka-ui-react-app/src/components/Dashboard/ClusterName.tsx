import React from 'react';
import { CellContext } from '@tanstack/react-table';
import { Tag } from 'components/common/Tag/Tag.styled';
import { Cluster } from 'generated-sources';

type ClusterNameProps = CellContext<Cluster, unknown>;

const ClusterName: React.FC<ClusterNameProps> = ({ row }) => {
  const { readOnly, name } = row.original;
  return (
    <>
      {readOnly && <Tag color="blue">readonly</Tag>}
      {name}
    </>
  );
};

export default ClusterName;
