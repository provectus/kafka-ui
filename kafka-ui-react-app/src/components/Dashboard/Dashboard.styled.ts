import styled from 'styled-components';

export const Toolbar = styled.div`
  padding: 8px 16px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  color: ${({ theme }) => theme.default.color.normal};
`;
