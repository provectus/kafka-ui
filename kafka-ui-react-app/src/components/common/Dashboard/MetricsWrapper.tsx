import React from 'react';
import cx from 'classnames';

interface Props {
  title?: string;
  wrapperClassName?: string;
  multiline?: boolean;
}

const MetricsWrapper: React.FC<Props> = ({
  title,
  children,
  wrapperClassName,
  multiline,
}) => {
  return (
    <div className={wrapperClassName}>
      {title && <h5 className="is-7 has-text-weight-medium mb-2">{title}</h5>}
      <div className="box">
        <div className={cx('level', multiline ? 'level-multiline' : '')}>
          {children}
        </div>
      </div>
    </div>
  );
};

export default MetricsWrapper;
