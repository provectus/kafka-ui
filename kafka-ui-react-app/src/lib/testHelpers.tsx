import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
