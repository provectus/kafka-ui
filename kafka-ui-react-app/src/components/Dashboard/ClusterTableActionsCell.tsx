import React from 'react';
import { Cluster } from 'generated-sources';
import { CellContext } from '@tanstack/react-table';
import { Button } from 'components/common/Button/Button';
import { clusterConfigPath } from 'lib/paths';

type Props = CellContext<Cluster, unknown>;

const ClusterTableActionsCell: React.FC<Props> = ({ row }) => {
  const { name } = row.original;
  return (
    <Button buttonType="secondary" buttonSize="S" to={clusterConfigPath(name)}>
      Configure
    </Button>
  );
};

export default ClusterTableActionsCell;
