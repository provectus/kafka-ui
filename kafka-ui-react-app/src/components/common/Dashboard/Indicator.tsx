import React from 'react';
import { styled } from 'lib/themedStyles';

interface Props {
  fetching?: boolean;
  label: React.ReactNode;
  title?: string;
  className?: string;
}

const Indicator: React.FC<Props> = ({
  label,
  title,
  fetching,
  className,
  children,
}) => {
  return (
    <div className={className}>
      <div title={title}>
        <p className="is-size-8">{label}</p>
        {fetching ? (
          <span className="icon has-text-grey-light">
            <i className="fas fa-spinner fa-pulse" />
          </span>
        ) : (
          children
        )}
      </div>
    </div>
  );
};

export default styled(Indicator)`
  background-color: white;
  height: 68px;
  width: 100%;
  max-width: 170px;
  min-width: 120px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  padding: 12px 16px;
  flex-grow: 1;

  box-shadow: 3px 3px 3px rgba(0, 0, 0, 0.08);
  margin: 0 0 3px 0;
`;
