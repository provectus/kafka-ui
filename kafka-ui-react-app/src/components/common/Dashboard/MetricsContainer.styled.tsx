import { styled } from 'lib/themedStyles';
import { Colors } from 'theme/theme';

export const MetricsContainerStyled = styled.div`
  padding: 1.5rem 1rem;
  background: ${Colors.neutral[5]};
  display: flex;
  grid-gap: 1rem;
  gap: 1rem;
  margin-bottom: 0.5rem !important;
  display: flex !important;
`;

export const MetricsLightText = styled.span`
  color: ${Colors.neutral[30]};
  font-size: 14px;
`;

export const MetricsRedText = styled.span`
  color: ${Colors.red[50]};
  font-size: 14px;
`;
