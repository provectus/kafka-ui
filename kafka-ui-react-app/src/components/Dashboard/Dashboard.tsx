import React from 'react';
import PageHeading from 'components/common/PageHeading/PageHeading';
import * as Metrics from 'components/common/Metrics';
import { Tag } from 'components/common/Tag/Tag.styled';
import Switch from 'components/common/Switch/Switch';
import { useClusters } from 'lib/hooks/api/clusters';
import { Cluster, ServerStatus } from 'generated-sources';
import { ColumnDef } from '@tanstack/react-table';
import Table, { SizeCell } from 'components/common/NewTable';
import useBoolean from 'lib/hooks/useBoolean';
import { Button } from 'components/common/Button/Button';
import { clusterNewConfigPath } from 'lib/paths';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';

import * as S from './Dashboard.styled';
import ClusterName from './ClusterName';
import ClusterTableActionsCell from './ClusterTableActionsCell';

const Dashboard: React.FC = () => {
  const clusters = useClusters();
  const { value: showOfflineOnly, toggle } = useBoolean(false);
  const appInfo = React.useContext(GlobalSettingsContext);

  const config = React.useMemo(() => {
    const clusterList = clusters.data || [];
    const offlineClusters = clusterList.filter(
      ({ status }) => status === ServerStatus.OFFLINE
    );
    return {
      list: showOfflineOnly ? offlineClusters : clusterList,
      online: clusterList.length - offlineClusters.length,
      offline: offlineClusters.length,
    };
  }, [clusters, showOfflineOnly]);

  const columns = React.useMemo<ColumnDef<Cluster>[]>(() => {
    const initialColumns: ColumnDef<Cluster>[] = [
      { header: 'Cluster name', accessorKey: 'name', cell: ClusterName },
      { header: 'Version', accessorKey: 'version' },
      { header: 'Brokers count', accessorKey: 'brokerCount' },
      { header: 'Partitions', accessorKey: 'onlinePartitionCount' },
      { header: 'Topics', accessorKey: 'topicCount' },
      { header: 'Production', accessorKey: 'bytesInPerSec', cell: SizeCell },
      { header: 'Consumption', accessorKey: 'bytesOutPerSec', cell: SizeCell },
    ];

    if (appInfo.hasDynamicConfig) {
      initialColumns.push({
        header: '',
        id: 'actions',
        cell: ClusterTableActionsCell,
      });
    }

    return initialColumns;
  }, []);

  return (
    <>
      <PageHeading text="Dashboard" />
      <Metrics.Wrapper>
        <Metrics.Section>
          <Metrics.Indicator label={<Tag color="green">Online</Tag>}>
            <span>{config.online || 0}</span>{' '}
            <Metrics.LightText>clusters</Metrics.LightText>
          </Metrics.Indicator>
          <Metrics.Indicator label={<Tag color="gray">Offline</Tag>}>
            <span>{config.offline || 0}</span>{' '}
            <Metrics.LightText>clusters</Metrics.LightText>
          </Metrics.Indicator>
        </Metrics.Section>
      </Metrics.Wrapper>
      <S.Toolbar>
        <div>
          <Switch
            name="switchRoundedDefault"
            checked={showOfflineOnly}
            onChange={toggle}
          />
          <label>Only offline clusters</label>
        </div>
        {appInfo.hasDynamicConfig && (
          <Button buttonType="primary" buttonSize="M" to={clusterNewConfigPath}>
            Configure new cluster
          </Button>
        )}
      </S.Toolbar>
      <Table
        columns={columns}
        data={config?.list}
        enableSorting
        emptyMessage={clusters.isFetched ? 'No clusters found' : 'Loading...'}
      />
    </>
  );
};

export default Dashboard;
