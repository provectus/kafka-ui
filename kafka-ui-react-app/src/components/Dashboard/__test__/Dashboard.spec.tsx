import React from 'react';
import Dashboard from 'components/Dashboard/Dashboard';
import { render } from 'lib/testHelpers';
import { screen } from '@testing-library/dom';

jest.mock('components/Dashboard/ClustersWidget/ClustersWidget', () => () => (
  <div>mock-ClustersWidget</div>
));

describe('Dashboard', () => {
  it('renders ClustersWidget', () => {
    render(<Dashboard />);
    expect(screen.getByText('Dashboard')).toBeInTheDocument();
    expect(screen.getByText('mock-ClustersWidget')).toBeInTheDocument();
  });
});
