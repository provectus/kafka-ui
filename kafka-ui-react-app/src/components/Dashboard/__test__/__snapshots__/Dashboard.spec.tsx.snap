// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard matches snapshot 1`] = `false`;
